### PR TITLE
[chai] Provide type definitions for chai.use

### LINKED ISSUE #29922: @types/chai No types for Chai's internal api (parameters of chai.use have type any)
- [x] I tried using the `@types/chai` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @jedmao @Bartvds @AGBrown @olivr70 @mwistrand @JoshuaKGoldberg @shaunluttin @Naktibalda @micksatana @ErikSchierboom

`Chai.ChaiStatic.use` is typed as `use(fn: (chai: any, utils: any) => void): ChaiStatic`.

The parameters `chai` and `utils` do have types, but they are part of the "private" API of chai, see for example `chai.Assertion` in [chaijs/chai](https://github.com/chaijs/chai/blob/master/lib/chai/assertion.js#L59). I'm trying to extend Chai, which means I would like to have the type of (amongst others) `chai.Assertion`.

Is there a reason why the internal API is not typed? This would be useful for Chai plugin developers.

### PR DESCRIPTION
**Pull request by 9999years**
Provide types for the following APIs used in plugin development, as requested in #29922:
- `chai.util`
- `chai.use()`
- `chai.Assertion`

Closes #29922.

## Version number

This PR includes breaking changes which will cause code that was incorrect (but
type-checked without errors) to no longer compile by aligning the type
definitions with Chai's actual public interface.

The only members with a (potentially) breaking change, however, is `chai.use`, which takes a `Chai.ChaiPlugin` rather than a `(chai: any, utils: any) => void` function.

This puts us in the uncomfortable position of needing to release a breaking change *while* not getting out of sync with Chai's major version. I'm not entirely sure what the best way to resolve this is, but I'm open to suggestions. One thing we could do is look for packages that depend on `@types/chai` and see if this change would break them.

## TypeScript version bumps

**Note:** Version bumps to TypeScript 3.0 for better tuple type matching; a lot of Chai's plugin APIs accept an array which is used as arguments to `chai.Assertion.prototype.assert`; that function has several optional parameters, so the last few elements of the tuple are optional. Therefore, the tuple is variable-length, which wasn't supported until TypeScript 3.0.

As a consequence, this PR bumps the minimum TypeScript version for a **lot** of packages. Sorry about that.

---

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://www.chaijs.com/api/plugins/
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header. **N/A**
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. **N/A:** Already exists.

### CODE REVIEW

File: types/chai/chai-tests.ts [hunk 1]
```diff
@@ -999,17 +999,91 @@ function _throw() {
...
-    });
-
-    const expect = chai
-        .use((_chai, util) => {
-        })
-        .use((_chai, util) => {
-        })
-        .expect;
-}
+                   // Modified from:
```
Comment (jedmao) [inline types/chai/chai-tests.ts]: Why is this indented so much?

**Review by jednano — APPROVED** [review]
Looks good, so long as the build passes.

File: types/chai/chai-tests.ts (hunk 1 above)
Comment (9999years) [reply types/chai/chai-tests.ts]: Unsure. Vim weirdness. Fixed, though.

**Review by jednano — APPROVED** [review]

**Comment by facetious** [comment]
What happened here?

`node_modules/@types/chai/index.d.ts:121:9 - error TS8020: JSDoc types can only be used inside documentation comments.
121         any?, // actual value
            ~~~~
node_modules/@types/chai/index.d.ts:122:9 - error TS8020: JSDoc types can only be used inside documentation comments.
122         boolean? // showDiff
            ~~~~~~~~
node_modules/@types/chai/index.d.ts:126:16 - error TS2370: A rest parameter must be of an array type.
126         assert(...args: AssertionArgs): void;`

This is with the new @types/chai@4.2.0.

**Comment by jimlindeman** [comment]
This change to use chai.chaiPlugin broke the prior backward compatibility with @types/chai@4.1.7. 
If try to go to @types/chai@4.2.0, it's not compatible with the typescript 2.9.x we're using and fails.  
@facetious see the output from "npm info @types/chai", that's why it failed for you.
```
dist-tags:
latest: 4.2.0  ts2.1: 4.1.7   ts2.3: 4.1.7   ts2.5: 4.1.7   ts2.7: 4.1.7   ts2.9: 4.1.7   ts3.1: 4.2.0   ts3.3: 4.2.0   ts3.5: 4.2.0   
ts2.0: 4.1.7   ts2.2: 4.1.7   ts2.4: 4.1.7   ts2.6: 4.1.7   ts2.8: 4.1.7   ts3.0: 4.2.0   ts3.2: 4.2.0   ts3.4: 4.2.0   ts3.6: 4.2.0   
```
Guess we'll just hardcode to @type/chai-as-packaged@7.1.1 for now.  Was there a reason this was a fix version bump and not at least a minor or major version bump?

**Comment by facetious** [comment]
Thanks, @jimlindeman. This was a typescript version issue, like you said. Guess I'll update to 3+.

**Comment by leonard-thieu** [comment]
> Was there a reason this was a fix version bump and not at least a minor or major version bump?

Version numbers for `@types` packages are aligned with the major and minor versions of the library package they represent.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library

**Comment by 9999years** [comment]
Yeah, I'm not the biggest fan of that policy (as I noted in the original PR description). That said, I'd love to help fix any packages this might've broken, because this situation is less than ideal to say the least. There's not a ton I can do about breaking compatibility with TypeScript@<3.0, though.

**Pull request merged**